Documentation/design-docs/first-class-structs.md

```diff
@@ -15,6 +15,261 @@ Primary Objectives
+
+### Struct “objects” as lvalues
+
+* The lhs of a struct assignment is a block node or local
+  * `GT_OBJ` nodes represent the “shape” info via a struct handle, along with the GC info
+    (location and type of GC references within the struct).
+    * These are used only to represent struct values that contain GC references.
+  * `GT_BLK` nodes represent struct types with no GC references, or opaque blocks of fixed size.
+    * These have no struct handle, resulting in some pessimization or even incorrect
+      code when the appropriate struct handle can't be determined.
+    * These never represent lvalues of structs that contain GC references.
+    * Proposed: Like `GT_OBJ`, these would have an index into the `ClassLayout` cache.
+  * `GT_STORE_OBJ` and `GT_STORE_BLK` have the same structure as `GT_OBJ` and `GT_BLK`, respectively
+    * `Data()` is op2
+  * `GT_DYN_BLK` and `GT_STORE_DYN_BLK` (GenTreeDynBlk extends GenTreeBlk)
+    * Additional child `gtDynamicSize`
+    * Note that these aren't really struct types; they represent dynamically sized blocks
+      of arbitrary data.
+  * For `GT_LCL_VAR` nodes, the shape information is obtained via the `LclVarDsc`
+
+### Struct “objects” as rvalues
+
+Structs only appear as rvalues in the following contexts:
+
+* On the RHS of an assignment
+  * The lhs provides the “shape” for an assignment. Note, however, that the LHS isn't always available
+    to optimizations, which has led to pessimization, e.g. 
+    [#23739 Block the hoisting of TYP_STRUCT rvalues in loop hoisting](https://github.com/dotnet/coreclr/pull/23739)
+
+* As a call argument
+  * In this context, it must be one of: `GT_OBJ`, `GT_LCL_VAR`, `GT_LCL_FLD` or `GT_FIELD_LIST`
+
+* As an operand to a hardware or SIMD intrinsic (for `TYP_SIMD*` only)
+  * In this case the struct handle is generally assumed to be unneeded, as it is captured (directly or
+    indirectly) in the `GT_SIMD` or `GT_HWIntrinsic` node.
+
+After morph, a struct-typed value on the RHS of assignment is one of:
+* `GT_IND`: in this case the LHS is expected to provide the struct handle
+  * Proposed: `GT_IND` would no longer be used for struct types
+* `GT_CALL`
+* `GT_RET_EXPR`
+* `GT_LCL_VAR`
+* `GT_LCL_FLD`
+  * Proposed: These would only be used for actual fields.
+* `GT_SIMD`
+* `GT_OBJ` nodes can also be used as rvalues when they are call arguments
+  * Proposed: `GT_OBJ` nodes can be used in any context where a struct rvalue or lvalue might occur,
+    except after morph when the struct is independently promoted.
+
+Struct IR Phase Transitions
+---------------------------
+
+There are three phases in the JIT that make changes to the representation of struct nodes and lclVars:
+
+* Importer
+  * Vector types are normalized to the appropriate `TYP_SIMD*` type. Other struct nodes have `TYP_STRUCT`.
+  * Proposed: all struct-valued nodes that are created with a class handle will retain an index into
```

Comment:
Small nit - the ClassLayout index is only needed when there are space constraints (e.g. there's no room in `LclFld` for a new pointer but we can fit in a `uint16_t` index. It can also be useful in VN. Otherwise using a ClassLayout pointer is preferrable.